train.py

```diff
@@ -0,0 +1,230 @@
+
+    # Save as an animated GIF
+    imgs[0].save(f"{chart_path}progress.gif", save_all=True, append_images=imgs[1:], loop=0, duration=500)
+
+
+def train_step(
+    generator: Generator,
+    discriminator: Discriminator,
+    train_loader: DataLoader,
+    gen_optimizer: torch.optim,
+    disc_optimizer: torch.optim,
+    criterion: nn.Module,
+    cfg: DictConfig,
+    fixed_noise: torch.Tensor,
+    epoch: int,
+) -> None:
+    generator.train()
+    discriminator.train()
+
+    def random_labels(batch_size, start, end, device):
+        return (end - start) * torch.rand(batch_size, device=device) + start
+
+    fig_list = []
+
+    progress_bar = tqdm(enumerate(train_loader), total=len(train_loader))
+    for batch_idx, batch in progress_bar:
+        real_data = batch[0].to(cfg.system.device)
+        batch_size = real_data.size(0)
+
+        real_labels = random_labels(batch_size, 0.8, 1, cfg.system.device)
+        fake_labels = random_labels(batch_size, 0.0, 0.1, cfg.system.device)
+
+        # train discriminator
+        discriminator.zero_grad()
+        label = real_labels
+        real_data = real_data.view(real_data.shape[0], 1, real_data.shape[1]).to(cfg.system.device)
+        disc_real_output = discriminator(real_data).view(-1)
+        discriminator_error_real = criterion(disc_real_output, label)
+        discriminator_error_real.backward()
+        D_x = disc_real_output.mean().item()  # Mean discriminator output for real data
+
+        # train generator
+        noise = torch.randn(batch_size, cfg.generator.nz, cfg.generator.output_channels, device=cfg.system.device)
+        fake_data = generator(noise)
+        label = fake_labels
+        disc_fake_output = discriminator(fake_data.detach()).view(-1)
+
+        discriminator_error_fake = criterion(disc_fake_output, label)
+        discriminator_error_fake.backward()
+        D_G_z1 = disc_fake_output.mean().item()  # Discriminator's average output when evaluating the fake data
+        discriminator_error = discriminator_error_real + discriminator_error_fake
+        disc_optimizer.step()
+        generator.zero_grad()
+
+        label = real_labels
+        disc_output_after_update = discriminator(fake_data).view(-1)
+        generator_error = criterion(disc_output_after_update, label)
+        generator_error.backward()
+        D_G_z2 = disc_output_after_update.mean().item()  # Discriminator's output after updating the generator, fake data
+        gen_optimizer.step()
+
+        # log to wandb
+        wandb.log(
+            {
+                "generator_error": generator_error.item(),
+                "discriminator_error": discriminator_error.item(),
+                "D_x": D_x,
+                "D_G_z1": D_G_z1,
+                "D_G_z2": D_G_z2,
+            }
+        )
+        if batch_idx % cfg.train.log_interval == 0:
+            fig = visualize_training(generator, fixed_noise, epoch, batch_idx, cfg.logger.chart_path)
+            fig_list.append(fig)
```

Comment:
```suggestion
        if batch_idx % cfg.train.log_interval == 0:
            wandb.log(
                {
                    "generator_error": generator_error.item(),
                    "discriminator_error": discriminator_error.item(),
                    "D_x": D_x,
                    "D_G_z1": D_G_z1,
                    "D_G_z2": D_G_z2,
                }
            )
        
            fig = visualize_training(generator, fixed_noise, epoch, batch_idx, cfg.logger.chart_path)
            fig_list.append(fig)
```

I don't think we need logis for every batch in wandb :)

Reply:
Changed, I set the `log_interval` to 100, feel free to change it.
Also added neat looking charts to wandb page, no more struggling with locally saved plots :tada:
![image](https://github.com/Nospoko/ecg-gan/assets/35274379/cdd9a329-9887-482a-902e-66fe97662ea3)